### ISSUE TITLE
tflint is slow with large number of provider instances

### ISSUE THREAD

**Issue by rquadling** [terraform-linters/tflint-ruleset-aws#1113]
The requirement: Extract and document various resource attributes from all of our accounts and regions. Things related to potential collisions / issues like VPC CIDR Blocks, open security groups, Load Balancer Listener Rule Priorities, etc. Mainly areas where we want to have a better overview of the whole setup. Look for inconsistencies, etc.

We have generators for the list of providers and the resource gathering elements.

```txt
$ cat generated_providers.tf  | grep provider | wc -l
      60
```

Terraform is happily gathering all the data and reporting it, all as required.

But our workflow includes TFLint as part of PreCommit, as well as TFLint in the repo pipelines. Make sure everything is as OK as possible before `terraform apply` takes place.

But when TFLint runs against this, we get a false error after about 2 minutes:
```txt
$ tflint
Failed to check ruleset; Cannot assume IAM Role; IAM Role (arn:aws:iam::4**********5:role/********Role) cannot be assumed.

There are a number of possible causes of this - the most common are:
  * The credentials used in order to assume the role are invalid
  * The credentials do not have appropriate permission to assume the role
  * The role ARN is not valid

Error: operation error STS: AssumeRole, https response error StatusCode: 0, RequestID: , request send failed, Post "https://sts.eu-west-3.amazonaws.com/": dial tcp: lookup sts.eu-west-3.amazonaws.com: no such host
```

If we tell TFLint to run with `-tflint --no-parallel-runners --max-workers=1`, everything works, but it takes significantly longer:
```txt
$ time tflint --no-parallel-runners --max-workers=1
2 issue(s) found:

generated_module_vpc_cidr_blocks.tf:676:1: Warning - output "cidr_blocks_by_account_region" should be moved from generated_module_vpc_cidr_blocks.tf to outputs.tf (terraform_standard_module_structure)
generated_module_vpc_cidr_blocks.tf:681:1: Warning - output "cidr_blocks_by_cidr_blocks" should be moved from generated_module_vpc_cidr_blocks.tf to outputs.tf (terraform_standard_module_structure)

real	39m14.298s
user	1m12.844s
sys	0m41.109s
```

I don't envisage the 2 errors being the cause of the problem.

Whilst running this, I grabbed the list of connections that were in play from MacOS Activity Monitor (straight select and copy):
```txt
192.168.88.43:55924->ec2-13-43-46-29.eu-west-2.compute.amazonaws.com:https
192.168.88.43:55925->ec2-13-43-46-29.eu-west-2.compute.amazonaws.com:https
192.168.88.43:55926->ec2-13-43-46-29.eu-west-2.compute.amazonaws.com:https
192.168.88.43:55927->ec2-13-43-46-29.eu-west-2.compute.amazonaws.com:https
192.168.88.43:55928->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:55929->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:55930->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55931->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55932->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55933->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55934->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55935->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55936->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55937->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55938->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55939->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55940->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:55941->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:55942->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55943->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55944->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:55945->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:55946->ec2-3-253-224-186.eu-west-1.compute.amazonaws.com:https
192.168.88.43:55947->ec2-3-253-224-186.eu-west-1.compute.amazonaws.com:https
192.168.88.43:55948->ec2-3-253-224-186.eu-west-1.compute.amazonaws.com:https
192.168.88.43:55949->ec2-3-253-224-186.eu-west-1.compute.amazonaws.com:https
192.168.88.43:55950->ec2-13-43-46-29.eu-west-2.compute.amazonaws.com:https
192.168.88.43:55951->ec2-13-43-46-29.eu-west-2.compute.amazonaws.com:https
192.168.88.43:55952->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:55953->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:55954->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55955->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55956->ec2-3-253-224-186.eu-west-1.compute.amazonaws.com:https
192.168.88.43:55957->ec2-3-253-224-186.eu-west-1.compute.amazonaws.com:https
192.168.88.43:55958->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55959->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55960->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55961->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55962->ec2-3-253-224-186.eu-west-1.compute.amazonaws.com:https
192.168.88.43:55963->ec2-3-253-224-186.eu-west-1.compute.amazonaws.com:https
192.168.88.43:55964->ec2-3-253-224-186.eu-west-1.compute.amazonaws.com:https
192.168.88.43:55965->ec2-3-253-224-186.eu-west-1.compute.amazonaws.com:https
192.168.88.43:55966->ec2-13-43-46-29.eu-west-2.compute.amazonaws.com:https
192.168.88.43:55967->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:55968->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55969->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55970->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:55971->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:55972->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55973->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55974->ec2-3-253-224-186.eu-west-1.compute.amazonaws.com:https
192.168.88.43:55975->ec2-3-253-224-186.eu-west-1.compute.amazonaws.com:https
192.168.88.43:55976->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:55977->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:55978->ec2-3-253-224-186.eu-west-1.compute.amazonaws.com:https
192.168.88.43:55979->ec2-3-253-224-186.eu-west-1.compute.amazonaws.com:https
192.168.88.43:55980->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:55981->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:55982->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:55983->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:55984->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55985->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55986->ec2-3-253-224-186.eu-west-1.compute.amazonaws.com:https
192.168.88.43:55987->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:55988->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:55989->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:55990->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:55991->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:55992->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:55993->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:55994->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55995->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:55996->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:55997->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:55998->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:55999->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56000->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56001->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56002->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56003->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56004->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:56005->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:56006->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:56007->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:56008->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56009->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56010->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56011->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56012->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56013->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56014->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:56015->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:56016->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56017->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56018->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:56019->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:56020->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56021->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56022->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56023->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56024->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56025->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56026->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56027->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56028->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56029->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56030->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56031->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56032->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56033->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56034->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:56035->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:56036->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56037->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56038->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:56039->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:56040->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56041->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56042->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56043->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56044->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:56045->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:56046->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:56047->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:56048->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56049->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56050->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56051->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56052->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:56053->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:56054->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56055->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56056->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56057->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56058->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:56059->ec2-13-217-78-225.compute-1.amazonaws.com:https
192.168.88.43:56060->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56061->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56062->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56063->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56064->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56065->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56066->ec2-13-39-220-255.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56067->ec2-13-39-225-156.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56068->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56069->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56070->ec2-13-39-225-156.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56071->ec2-13-39-225-156.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56072->67.220.243.61:https
192.168.88.43:56073->67.220.243.61:https
192.168.88.43:56074->ec2-13-39-225-156.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56075->ec2-13-39-225-156.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56076->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56077->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56078->ec2-13-39-225-156.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56079->ec2-13-39-225-156.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56080->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56081->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56082->ec2-13-39-225-156.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56083->ec2-13-39-225-156.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56084->ec2-13-39-225-156.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56085->ec2-13-39-225-156.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56086->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56087->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56088->ec2-13-39-225-156.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56089->ec2-13-39-225-156.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56090->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56091->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56092->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56093->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56094->ec2-13-39-225-156.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56095->ec2-13-39-225-156.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56096->67.220.243.61:https
192.168.88.43:56097->67.220.243.61:https
192.168.88.43:56098->ec2-13-39-225-156.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56099->ec2-13-39-225-156.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56100->67.220.243.61:https
192.168.88.43:56101->67.220.243.61:https
192.168.88.43:56102->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56103->ec2-13-43-48-181.eu-west-2.compute.amazonaws.com:https
192.168.88.43:56104->ec2-13-39-225-156.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56105->ec2-13-39-225-156.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56106->67.220.243.61:https
192.168.88.43:56107->67.220.243.61:https
192.168.88.43:56108->ec2-13-39-225-156.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56109->ec2-13-39-225-156.eu-west-3.compute.amazonaws.com:https
192.168.88.43:56110->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56111->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56112->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
192.168.88.43:56113->ec2-3-253-222-165.eu-west-1.compute.amazonaws.com:https
```

I suspect there is some limit I'm hitting. We do have other repos with less than 10 providers and not had an issue.

**Comment by rquadling** [comment]
We fixed the `terraform_standard_module_structure` issue and re-ran the TFLint:

```txt
 $ time tflint --no-parallel-runners --max-workers=2

real	42m21.357s
user	1m23.372s
sys	0m43.494s
```
As expected (I hope), increasing the `--max-workers` whilst still having `--no-parallel-runners` didn't make any difference.

With no `--no-parallel-runners`:
```txt
$ time tflint --max-workers=2
Failed to check ruleset; validating provider credentials: retrieving caller identity from STS: operation error STS: GetCallerIdentity, https response error StatusCode: 0, RequestID: , request send failed, Post "https://sts.eu-west-1.amazonaws.com/": dial tcp: lookup sts.eu-west-1.amazonaws.com: no such host; 

real	0m55.581s
user	0m26.675s
sys	0m4.131s
```

And even with `--no-parallel-runners --max-workers=1`:
```txt
$ time tflint --max-workers=1
Failed to check ruleset; validating provider credentials: retrieving caller identity from STS: operation error STS: GetCallerIdentity, https response error StatusCode: 0, RequestID: , request send failed, Post "https://sts.eu-west-1.amazonaws.com/": dial tcp: lookup sts.eu-west-1.amazonaws.com: no such host; 

real	0m47.278s
user	0m24.697s
sys	0m3.322s
```

At this stage, we have to NOT run TFLint on this repo. Admittedly this really isn't a major issue for us, but possible there's something in the ruleset code that could be improved as whilst our use case is from generated code, for others it may be a real multi-account/region setup and a real issue.

If there is anyone with the knowledge/skills to provide PRs for this repo, we're happy to help test things out.

**Comment by rquadling** [comment]
It seems that the `deep_check` is triggering the issue.
```txt
$ time tflint

real	0m1.432s
user	0m5.058s
sys	0m2.195s
```

**Comment by rquadling** [comment]
Hi. What is the timeframe for a release on this please?
